src/module-server/plugins/npm-plugin.ts

```diff
@@ -166,58 +197,60 @@ let npmCache: RollupCache | undefined;
  */
-const bundleNpmModule = async (mod: string, optimize: boolean) => {
+const bundleNpmModule = async (mod: string, id: string, optimize: boolean) => {
+  let namedExports: string[] = [];
+  if (dynamicCJSModules.has(id)) {
+    let isValidCJS = true;
+    try {
+      const text = await fs.readFile(mod, 'utf8');
+      // Goal: Determine if it is ESM or CJS.
+      // Try to parse it with cjs-module-lexer, if it fails, assume it is ESM
+      // eslint-disable-next-line @cloudfour/typescript-eslint/await-thenable
+      await parse(text);
+    } catch {
+      isValidCJS = false;
+    }
+
+    if (isValidCJS) {
+      const require = createRequire(import.meta.url);
+      // eslint-disable-next-line @cloudfour/typescript-eslint/no-var-requires
+      const imported = require(mod);
+      if (typeof imported === 'object' && !imported.__esModule)
+        namedExports = Object.keys(imported);
+    }
+  }
+
+  const virtualEntry = '\0virtualEntry';
+  const hasSyntheticNamedExports = namedExports.length > 0;
   const bundle = await rollup({
-    input: mod,
+    input: hasSyntheticNamedExports ? virtualEntry : mod,
     cache: npmCache,
     shimMissingExports: true,
     treeshake: true,
     preserveEntrySignatures: 'allow-extension',
     plugins: [
-      {
-        // This plugin fixes cases of module.exports = require('...')
-        // By default, the named exports from the required module are not generated
-        // This plugin detects those exports,
-        // and makes it so that @rollup/plugin-commonjs can see them and turn them into ES exports (via syntheticNamedExports)
-        // This edge case happens in React, so it was necessary to fix it.
-        name: 'cjs-module-lexer',
-        async transform(code, id) {
-          if (id.startsWith('\0')) return;
-          const out = new MagicString(code);
-          const re =
-            /(^|[\s;])module\.exports\s*=\s*require\(["']([^"']*)["']\)($|[\s;])/g;
-          let match;
-          while ((match = re.exec(code))) {
-            const [, leadingWhitespace, moduleName, trailingWhitespace] = match;
-
-            const resolved = await this.resolve(moduleName, id);
-            if (!resolved || resolved.external) return;
-
-            try {
-              const text = await fs.readFile(resolved.id, 'utf8');
-              // eslint-disable-next-line @cloudfour/typescript-eslint/await-thenable
-              const parsed = await parse(text);
-              let replacement = '';
-              for (const exportName of parsed.exports) {
-                replacement += `\nmodule.exports.${exportName} = require("${moduleName}").${exportName}`;
-              }
-
-              out.overwrite(
-                match.index,
-                re.lastIndex,
-                leadingWhitespace + replacement + trailingWhitespace,
-              );
-            } catch {
-              return;
+      namedExports.length > 0 &&
```

Comment:
Nit: this might be easier to read if you moved this logic outside of this config object.

Pseudo-code:

```js
const plugins = [];

if (namedExports.length > 0) {
  plugins.push({/* ... */});
}

const bundle = await rollup({
  /* ... */
  plugins,
  /* ... */
})
```

This is subjective though. I know the `&&` pattern is used pretty extensively in JSX so may be more familiar to other devs besides me.

Reply:
Yeah I think it's pretty common in config arrays like this (babel plugins, rollup plugins, etc.)